Recast/PlayerViewController.swift

```diff
@@ -8,59 +8,344 @@
+
+        let alertMessage = message ?? NSLocalizedString("error.default.description", comment: "Default error message when no NSError provided")
+
+        let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
+
+        let alertActionTitle = NSLocalizedString("alert.error.actions.OK", comment: "OK on error alert")
+        let alertAction = UIAlertAction(title: alertActionTitle, style: .default, handler: nil)
+
+        alert.addAction(alertAction)
+
+        present(alert, animated: true, completion: nil)
+    }
+
+    // MARK: Custom Play options
+
+    func play(_ episode: Episode) {
+        current = episode
+        guard let encl = current?.enclosure, let url = encl.url else {
+            // TODO: handle error
+            return
+        }
+        // TODO: update for downloaded episodes
+        let asset = AVAsset(url: url) // Use Assets for trimming later
+        let item = AVPlayerItem(asset: asset)
+
+        if #available(iOS 10, *) {
+            player.automaticallyWaitsToMinimizeStalling = false
+        }
+
+        player.pause()
+        player.replaceCurrentItem(with: item)
+        player.play()
+
+        updateNowPlayingInfo()
+        updatePlayerUI()
+    }
+
+    // TODO: Queueing
+    func queue(_ episode: Episode, at index: Int? = nil) {
+        if queue.isEmpty {
+            play(episode)
+            return
+        }
+        if let i = index {
+            queue.insert(episode, at: i)
+        } else {
+            queue.append(episode)
+        }
+    }
+
+    func updateDuration(_ newDuration: CMTime) {
+        let hasValidDuration = newDuration.isNumeric && newDuration.value != 0
+        let newDurationSeconds = hasValidDuration ? CMTimeGetSeconds(newDuration) : 0.0
+        let currentTime = hasValidDuration ? Float(CMTimeGetSeconds(player.currentTime())) : 0.0
+
+        controlsView.timeSlider.maximumValue = Float(newDurationSeconds)
+        controlsView.timeSlider.value = currentTime
+        controlsView.timeSlider.isContinuous = true
+        controlsView.rewindButton.isEnabled = hasValidDuration
+        controlsView.playPauseButton.isEnabled = hasValidDuration
+        controlsView.forwardButton.isEnabled = hasValidDuration
+        controlsView.timeSlider.isEnabled = hasValidDuration
+        controlsView.leftTimeLabel.isEnabled = hasValidDuration
+        controlsView.leftTimeLabel.text = createTimeString(time: currentTime)
+        controlsView.rightTimeLabel.isEnabled = hasValidDuration
+        controlsView.rightTimeLabel.text = createTimeString(time: Float(newDurationSeconds))
+    }
+
+    func updatePlayerUI() {
+
+    }
+
+    func updateNowPlayingInfo() {
+        if let imageUrl = current?.podcast?.artworkUrl600, episodeImageView != nil {
+            episodeImageView.kf.setImage(with: imageUrl)
+        }
+    }
+
+    // MARK: Convenience
+
+    func createTimeString(time: Float) -> String {
```

Comment:
move to utilities?

Reply:
It's only used here, ideally it's more efficient to leave here.